components/brave_news/browser/signal_calculator.cc

```diff
@@ -0,0 +1,192 @@
+
+  base::flat_map<std::string, std::vector<std::string>> publisher_visits;
+  base::flat_map<std::string, std::vector<std::string>> channel_visits;
+
+  for (auto& [publisher_id, publisher] : publishers) {
+    auto host = publisher->site_url.host();
+    auto history_it = origin_visits.find(host);
+    if (history_it == origin_visits.end()) {
+      publisher_visits[publisher_id] = {};
+      continue;
+    }
+
+    base::ranges::copy(history_it->second,
+                       std::back_inserter(publisher_visits[publisher_id]));
+    total_publisher_visits += history_it->second.size();
+
+    for (const auto& locale_info : publisher->locales) {
+      if (locale_info->locale != locale) {
+        continue;
+      }
+
+      for (const auto& channel : locale_info->channels) {
+        total_channel_visits += history_it->second.size();
+        base::ranges::copy(history_it->second,
+                           std::back_inserter(channel_visits[channel]));
+      }
+      break;
+    }
+  }
+
+  Signals signals;
+
+  // Add article signals
+  for (const auto& article : articles) {
+    const auto& publisher = publishers.at(article->publisher_id);
+
+    bool channel_subscribed = false;
+    for (const auto& locale_info : publisher->locales) {
+      for (const auto& channel : locale_info->channels) {
+        if (channels_controller_->GetChannelSubscribed(locale_info->locale,
+                                                       channel)) {
+          channel_subscribed = true;
+          break;
+        }
+      }
+    }
+    auto& visits = publisher_visits.at(article->publisher_id);
+    signals[article->url.spec()] = {
+        .subscribed = IsPublisherSubscribed(publisher->user_enabled_status,
+                                            channel_subscribed),
+        .visit_weight =
+            visits.size() / static_cast<double>(total_publisher_visits),
+        .pop_recency = GetPopRecency(article)};
+  }
+
+  // Add publisher signals
+  for (const auto& it : publishers) {
+    bool channel_subscribed = false;
+    for (const auto& locale_info : it.second->locales) {
+      for (const auto& channel : locale_info->channels) {
+        if (channels_controller_->GetChannelSubscribed(locale_info->locale,
+                                                       channel)) {
+          channel_subscribed = true;
+          break;
+        }
+      }
+    }
+    auto& visits = publisher_visits.at(it.first);
+    signals[it.first] = {
+        .subscribed = IsPublisherSubscribed(it.second->user_enabled_status,
+                                            channel_subscribed),
+        .visit_weight =
+            visits.size() / static_cast<double>(total_publisher_visits),
+        .pop_recency = 0};
+  }
+
+  // Add channel signals
+  for (const auto& channel : channels) {
+    auto it = channel_visits.find(channel.first);
+    std::vector<std::string> visits =
+        it == channel_visits.end() ? std::vector<std::string>{} : it->second;
```

Comment:
I guess this would result in copying the entire vector, but the only thing we need seems to be the size of it. How about return 0 when `it == channel_visits.end()`?

```suggestion
    size_t visits_size =
        it == channel_visits.end() ? 0u : it->second.size();
```

Reply:
Nice catch! Yeah, originally I was adding all the visited pages to the signal, but I think it was overkill.